finetuning-llama-7b-on-l4/README.md

```diff
@@ -0,0 +1,381 @@
+              name: l4-demo
+              key: HF_TOKEN
+        volumeMounts:
+        - name: gcs-fuse-csi-ephemeral
+          mountPath: /gcs-mount
+      serviceAccountName: l4-demo
+      volumes:
+      - name: gcs-fuse-csi-ephemeral
+        csi:
+          driver: gcsfuse.csi.storage.gke.io
+          volumeAttributes:
+            bucketName: $BUCKET_NAME
+            mountOptions: "implicit-dirs"
+EOF
+```
+
+Give it a minute to start running, once up you can watch the logs of the job by running:
+```bash
+kubectl logs -f -l job-name=model-loader
+```
+
+Once the job has finished you can verify the model has been downloaded by running:
+```bash
+gcloud storage ls -l gs://$BUCKET_NAME/llama2-7b/
+```
+
+Let’s write our finetuning job code by using the HuggingFace library for training.
+
+Create a file named `finetune.py` with the following content:
+```python
+from pathlib import Path
+
+from datasets import load_dataset, concatenate_datasets
+import transformers
+from transformers import AutoTokenizer, AutoModelForCausalLM, TrainingArguments
+from peft import get_peft_model, LoraConfig, prepare_model_for_kbit_training
+import torch
+
+# /gcs-mount will mount the GCS bucket created earlier
+model_path = "/gcs-mount/llama2-7b"
+finetuned_model_path = "/gcs-mount/llama2-7b-american-stories"
+
+tokenizer = AutoTokenizer.from_pretrained(model_path, local_files_only=True)
+model = AutoModelForCausalLM.from_pretrained(
+            model_path, torch_dtype=torch.float16, device_map="auto", trust_remote_code=True)
+
+dataset = load_dataset("dell-research-harvard/AmericanStories",
+    "subset_years",
+    year_list=["1809", "1810", "1811", "1812", "1813", "1814", "1815"]
+)
+dataset = concatenate_datasets(dataset.values())
+
+if tokenizer.pad_token is None:
+    tokenizer.add_special_tokens({'pad_token': '[PAD]'})
+    model.resize_token_embeddings(len(tokenizer))
+
+data = dataset.map(lambda x: tokenizer(
+    x["article"], padding='max_length', truncation=True))
+
+lora_config2 = LoraConfig(
+ r=16,
+ lora_alpha=32,
+ lora_dropout=0.05,
+ bias="none",
+ task_type="CAUSAL_LM"
+)
+
+model = prepare_model_for_kbit_training(model)
+
+# add LoRA adaptor
+model = get_peft_model(model, lora_config2)
+model.print_trainable_parameters()
+
+training_args = TrainingArguments(
+        per_device_train_batch_size=1,
+        gradient_accumulation_steps=4,
+        warmup_steps=2,
+        num_train_epochs=1,
+        learning_rate=2e-4,
+        fp16=True,
+        logging_steps=1,
+        output_dir=finetuned_model_path,
+        optim="paged_adamw_32bit",
+)
+
+trainer = transformers.Trainer(
+    model=model,
+    train_dataset=data,
+    args=training_args,
+    data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False),
+)
+model.config.use_cache = False  # silence the warnings. Please re-enable for inference!
+
+trainer.train()
+
+# Merge the fine tuned layer with the base model and save it
+model = model.merge_and_unload()
```

Comment:
Per earlier comment, can this step be skipped to optimize storage?

Reply:
Yes it could but I rather not. I've had too much pain dealing with LoRA layers that are separated from the base model.

I will add that this step can be removed if you want to only save the LoRA layer. I've also added an explanation below in the tutorial.